cpp/src/fil/fil.cu

```diff
@@ -636,6 +637,89 @@ int tree2fil_sparse(std::vector<fil_node_t>& nodes, int root,
   return root;
 }
 
+struct level_entry {
+  int n_branches, n_leaves;
+};
+typedef std::pair<int, int> pair_t;
+using std::tie;
+// hist has branch and leaf count given depth
+template <typename T, typename L>
+inline void tree_depth_hist(const tl::Tree<T, L>& tree,
+                            std::unordered_map<int, level_entry>& hist) {
+  std::stack<pair_t> stack;  // {tl_id, depth}
+  stack.push({tree_root(tree), 0});
+  while (!stack.empty()) {
+    int node_id, depth;
+    tie(node_id, depth) = stack.top();
+    stack.pop();
+    while (!tree.IsLeaf(node_id)) {
+      hist[depth].n_branches++;
+      stack.push({tree.LeftChild(node_id), depth + 1});
+      node_id = tree.RightChild(node_id);
+      depth++;
+    }
+    hist[depth].n_leaves++;
+  }
+}
+
+// implements https://tools.ietf.org/html/draft-eastlake-fnv-17.html
+// algorithm is public domain without export controls and no patents or rights to patent
+auto fowler_noll_vo_fingerprint64(int* begin, int* end) {
+  return std::accumulate(begin, end, 14695981039346656037lu,
+                         [](auto& fingerprint, int& x) {
+                           return (fingerprint * 1099511628211lu) ^ x;
+                         });
+}
+
+template <typename T, typename L>
+std::stringstream depth_hist_and_max(const tl::ModelImpl<T, L>& model) {
+  std::unordered_map<int, level_entry> depth_hist;
+  for (const auto& tree : model.trees) tree_depth_hist(tree, depth_hist);
+  struct hist_entry {
+    int level, n_branches, n_leaves;
+  };
+  std::vector<hist_entry> vec_hist;
+  for (auto p : depth_hist) {
+    vec_hist.emplace_back(
+      hist_entry{p.first, p.second.n_branches, p.second.n_leaves});
+  }
+  std::sort(vec_hist.begin(), vec_hist.end(),
+            [](auto a, auto b) { return a.level < b.level; });
+
+  int min_depth = -1, leaves_times_depth = 0, total_branches = 0,
+      total_leaves = 0;
+  std::stringstream forest_shape;
+  forest_shape << "Depth histogram:\n"
+               << "depth branches leaves   nodes\n";
+  char line[100];
+  for (hist_entry e : vec_hist) {
+    sprintf(line, "%5d %8d %6d %7d\n", e.level, e.n_branches, e.n_leaves,
```

Comment:
If you're writing into an object from `<iostream>`, it's better to use IO manipulators: https://en.cppreference.com/w/cpp/io/manip.